### PR TITLE
Sync environment specific branches with master

### PR DESCRIPTION
**Pull request by vijayrajagopalan-hmcts**
Resolves # . (This is applicable only if this pull request relates to a GitHub issue, delete the line otherwise)

Notes:

Syncs environment specific branches such as demo, ithc etc with master

Executes sync as a last stage in the master. uil

Added to Readme

Tested with am-org-role-mapping-batch-service

Jenkinsfile https://github.com/hmcts/am-org-role-mapping-batch-service/blob/master/Jenkinsfile_CNP

Build here https://build.platform.hmcts.net/job/HMCTS_AM/job/am-org-role-mapping-batch-service/job/master/23/

### CODE REVIEW

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                    git remote set-url origin $(git config remote.origin.url | sed "s/github.com/${USER_NAME}:${BEARER_TOKEN}@github.com/g")
+
+                '''
+
+                for (branch in config.branchesToSyncWithMaster) {
+                    println('Syncing branch - ' + branch)
+
+                    try {
+                        sh '''git fetch origin '''+branch+''':'''+branch
+                        sh '''git push --force origin HEAD:'''+branch
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:35]: use multiline string for interpolation of strings
```suggestion
                        sh """
                         git fetch origin ${branch}:${branch}
                         git push --force origin HEAD:${branch}
                        """
```

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            
+            withCredentials([usernamePassword(credentialsId: credentialsId, passwordVariable: 'BEARER_TOKEN', usernameVariable: 'USER_NAME')]) {
+                sh '''
+                    set -e
+                    git remote set-url origin $(git config remote.origin.url | sed "s/github.com/${USER_NAME}:${BEARER_TOKEN}@github.com/g")
+
+                '''
+
+                for (branch in config.branchesToSyncWithMaster) {
+                    println('Syncing branch - ' + branch)
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:31]: println goes to the server log, you'll want echo to go to the jenkins log

```suggestion
                    echo "Syncing branch - ${branch}"
```

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import uk.gov.hmcts.contino.AppPipelineConfig
+
+def call(params) {
+    AppPipelineConfig config = params.appPipelineConfig
+    def product = params.product
+    def credentialsId = env.GIT_CREDENTIALS_ID
+
+    stageWithAgent("Sync Branches with Master", product) {
+        if (!config.branchesToSyncWithMaster.isEmpty()) {
+            // withCredentials([this.steps.usernamePassword(credentialsId: 'jenkins-github-hmcts-api-token', usernameVariable: 'USER_NAME', passwordVariable: 'BEARER_TOKEN')]) {
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:21]:
```suggestion
```

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+                    try {
+                        sh """
+                         git fetch origin ${branch}:${branch}
+                         git push --force origin HEAD:${branch}
+                        """
+                        
+                    } catch (err) {
+                        println("Failed to update $branch branch.")
+                        println(err.getMessage())
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:41]:
```suggestion
                        echo "Failed to update $branch branch."
                        echo err.getMessage()
```

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                        throw err
+                    }
+
+                    println('Sync completed for - ' + branch)
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:45]:
```suggestion
                    echo "Sync completed for - ${branch}"
```

**Review by timja — DISMISSED** [review]

**Review by timja — DISMISSED** [review]

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+                    try {
+                        sh """
+                         git fetch origin ${branch}:${branch}
+                         git push --force origin HEAD:${branch}
+                        """
+                        
+                    } catch (err) {
+                        echo "Failed to update $branch branch."
+                        echo err.getMessage()
+                        throw err
```
Comment (adusumillipraveen) [inline vars/sectionSyncBranchesWithMaster.groovy:40]: Let's make this a silent failure and add a note in README.

Comment (vijayrajagopalan-hmcts) [reply vars/sectionSyncBranchesWithMaster.groovy]: Removed throw statement

**Review by timja — APPROVED** [review]

**Review by adusumillipraveen — APPROVED** [review]

**Pull request merged**